File: ui/src/api/hooks/assets.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import {
  commands,
  type ManualAsset, type NewManualAsset, type ManualAssetPatch,
  type DebtPayoffResult,
} from "../client";
import { unwrap } from "../client";
import { isBackendAvailable } from "../../utils/runtime";
import { invalidateDomains } from "../invalidation";

export function useManualAssets() {
  return useQuery<ManualAsset[]>({
    queryKey: ["manual-assets"],
    queryFn: async () => {
      return unwrap(commands.listManualAssets());
    },
    enabled: isBackendAvailable(),
  });
}

export function useCreateManualAsset() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (input: NewManualAsset) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      return unwrap(commands.createManualAsset(input));
    },
    onSuccess: () => {
      invalidateDomains(qc, "manualAssets");
    },
  });
}

export function useUpdateManualAsset() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ id, patch }: { id: string; patch: ManualAssetPatch }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      return unwrap(commands.updateManualAsset(id, patch));
    },
    onSuccess: () => {
      invalidateDomains(qc, "manualAssets");
    },
  });
}

export function useDeleteManualAsset() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      await unwrap(commands.deleteManualAsset(id));
    },
    onSuccess: () => {
      invalidateDomains(qc, "manualAssets");
    },
  });
}

export function useDebtPayoff(extraMonthlyCents: number) {
  return useQuery<DebtPayoffResult[]>({
    queryKey: ["debt-payoff", extraMonthlyCents],
    queryFn: async () => {
      return unwrap(commands.computeDebtPayoff(extraMonthlyCents));
    },
    enabled: isBackendAvailable(),
  });
}

export function useUncelebratedMilestones() {
  return useQuery<number[]>({
    queryKey: ["networth-milestones"],
    queryFn: async () => {
      return unwrap(commands.getUncelebratedMilestones());
    },
    staleTime: Infinity,
    enabled: isBackendAvailable(),
  });
}

File: crates/finsight-app/src/commands/assets.rs
use crate::error::AppResult;
use crate::AppState;
use finsight_core::models::{ManualAsset, ManualAssetPatch, NetWorthPoint, NewManualAsset};

pub use finsight_api::commands::assets::{
    DebtPayoffMonth, DebtPayoffResult, DebtPayoffSummary,
};

#[tauri::command]
#[specta::specta]
pub async fn list_manual_assets(state: tauri::State<'_, AppState>) -> AppResult<Vec<ManualAsset>> {
    finsight_api::commands::assets::list_manual_assets(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn create_manual_asset(
    state: tauri::State<'_, AppState>,
    input: NewManualAsset,
) -> AppResult<ManualAsset> {
    finsight_api::commands::assets::create_manual_asset(&state.api, input).await
}

#[tauri::command]
#[specta::specta]
pub async fn update_manual_asset(
    state: tauri::State<'_, AppState>,
    id: String,
    patch: ManualAssetPatch,
) -> AppResult<ManualAsset> {
    finsight_api::commands::assets::update_manual_asset(&state.api, id, patch).await
}

#[tauri::command]
#[specta::specta]
pub async fn delete_manual_asset(state: tauri::State<'_, AppState>, id: String) -> AppResult<()> {
    finsight_api::commands::assets::delete_manual_asset(&state.api, id).await
}

#[tauri::command]
#[specta::specta]
pub async fn record_net_worth_snapshot(state: tauri::State<'_, AppState>) -> AppResult<()> {
    finsight_api::commands::assets::record_net_worth_snapshot(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn list_net_worth_history(
    state: tauri::State<'_, AppState>,
    days: u32,
) -> AppResult<Vec<NetWorthPoint>> {
    finsight_api::commands::assets::list_net_worth_history(&state.api, days).await
}

#[tauri::command]
#[specta::specta]
pub async fn compute_debt_payoff(
    state: tauri::State<'_, AppState>,
    extra_monthly_cents: i64,
) -> AppResult<Vec<DebtPayoffResult>> {
    finsight_api::commands::assets::compute_debt_payoff(&state.api, extra_monthly_cents).await
}

#[tauri::command]
#[specta::specta]
pub async fn get_uncelebrated_milestones(state: tauri::State<'_, AppState>) -> AppResult<Vec<i64>> {
    finsight_api::commands::assets::get_uncelebrated_milestones(&state.api).await
}
